### PR TITLE
source_code_interpretation/utils/autoanchor_py

### PR DESCRIPTION
**Pull request by ccssu**

### CODE REVIEW

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+欢迎star [one-yolov5项目](https://github.com/Oneflow-Inc/one-yolov5) 获取<a href="https://github.com/Oneflow-Inc/one-yolov5/tags" target="blank" >最新的动态。</a>
+
+
+<a href="https://github.com/Oneflow-Inc/one-yolov5/issues/new"  target="blank"  >如果您有问题，欢迎在仓库给我们提出宝贵的意见。🌟🌟🌟</a>
+
+<a href="https://github.com/Oneflow-Inc/one-yolov5" target="blank" >
+如果对您有帮助，欢迎来给我Star呀😊~  </a>
+
+## 摘要
+&emsp;**维度聚类**（Dimension Clusters）。当把锚框与YOLO一起使用时，我们会遇到两个问题。 首先是框的尺寸是手工挑选的。虽然网络可以通过学习适当地调整方框，但是如果我们从一开始就为网络选择更好的先验框，就可以让网络更容易学习到更好的检测结果。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:19]:
```suggestion
&emsp;**维度聚类**（Dimension Clusters）。当把锚框 (Anchor) 与YOLO一起使用时，我们会遇到两个问题。 首先锚框的尺寸是手工挑选的。虽然网络可以通过学习适当地调整锚框，但是如果我们从一开始就为网络选择更好的锚框，就可以让网络更容易学习到更好的检测结果。
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+![image.png](autoanchor_imgs/picture_00.png)
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:21]: 超链接

File: docs/source_code_interpretation/utils/autoanchor_py.md (hunk 1 above)
Comment (BBuf) [reply docs/source_code_interpretation/utils/autoanchor_py.md:19]: 或者本文统一叫先验框也可以。

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+图1：$VOC$ 和  $COCO$  上的聚类框尺寸。我们在边界框的维上运行 $k-means$ 聚类，以获得我们模型的良好先验。左图显示了我们通过k的各种选择获得的平均 $IOU$ 。我们发现 $k = 5$ 为召回与模型的复杂性提供了良好的折中。右图显示了 $VOC$ 和 $COCO$ 的相对质心。这两种方案都喜欢更稀疏的，更高的框，并且 $COCO$ 的尺寸的变化比 $VOC$ 更大。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:23]:
```suggestion
图1：$VOC 2007$ 和  $COCO$  上的聚类框尺寸。我们在边界框的维上运行 $k-means$ 聚类，以获得YOLO模型的良好先验框。左图显示了我们通过 k 的各种选择获得的平均 $IOU$ 。我们发现 $k = 5$ 为 **召回率** 与 **模型的复杂度** 提供了良好的折中。右图显示了 $VOC$ 和 $COCO$ 的相对质心。这两种方案都喜欢更稀疏的，更高的框，并且 $COCO$ 数据集的先验框尺寸的变化比 $VOC$ 更大。
```

相对质心需要添加解释

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+<a href="https://github.com/Oneflow-Inc/one-yolov5" target="blank" >
+如果对您有帮助，欢迎来给我Star呀😊~  </a>
+
+
+源码解读： [utils/autoanchor.py](https://github.com/Oneflow-Inc/one-yolov5/blob/main/utils/autoanchor.py)
+
+## 摘要
+&emsp;**维度聚类**（Dimension Clusters）。当把 YOLO 结合 $anchor \ boxes$ 使用时，我们会遇到两个问题： 首先 $anchor \ boxes$ 的尺寸是手工挑选的。虽然网络可以通过学习适当地调整$anchor \ boxes$ 形状，但是如果我们从一开始就为网络选择更好的 $anchor \ boxes$ ，就可以让网络更容易学习到并获得更好的检测结果。
+
+![image](https://user-images.githubusercontent.com/109639975/199901435-76986df9-cc7b-4eac-97f1-fc905ed3d8d7.png)
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:15]: 链接换了之后图可以删除了

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+&emsp;我们不用手工选择 $anchor \ boxes$，而是在训练集的边界框上的维度上运行 k-means聚类算法，自动找到良好的 $anchor \ boxes$ 。 如果我们使用具有欧几里得距离的标准 $k-means$ ，那么较大的框比较小的框产生更多的误差。 然而，我们真正想要的是独立于框的大小的，能获得良好的 $IoU$ 分数的 $anchor \ boxes$ 。 因此对于距离度量我们使用:
+
+<center>
+
+$d(\text { box, centroid }) = 1-\operatorname{IoU}(\text { box }, \text { centroid })$
+
+</center>
+
+&emsp;我们用不同的 $k$ 值运行 $k-means$ 算法，并绘制最接近质心的平均 $Avg \ IoU$（见图1）。为了在模型复杂度和高召回率之间的良好折中，我们选择 $k = 5$ （*也就是5种anchor boxs*）簇的相对中心 与手工选取的 $anchor \ boxes$ 显着不同，它有更少的短且宽的框，并且有更多既长又窄的框。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:28]:
```suggestion
&emsp;我们用不同的 $k$ 值运行 $k-means$ 算法，并绘制最接近质心的平均 $Avg \ IoU$（见图1）。为了在模型复杂度和高召回率之间的良好折中，我们选择 $k = 5$ （*也就是5种anchor boxes*）簇的相对中心 与手工选取的 $anchor \ boxes$ 显着不同，它有更少的短且宽的框，并且有更多既长又窄的框。
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+
+
+&emsp;表1中，我们将聚类策略的 $anchor \ boxes$ 中心数和手工选取的 $anchor \ boxes$ 在最接近的 $Avg \ IoU$ 上进行比较。仅5种 $anchor \ boxes$ 中心的 $Avg \ IoU$ 为61.0，其性能类似于9个$anchor \ boxes$ 的60.9 (*即Avg IoU已经达到了Faster RCNN的水平*)。 而且使用9种 $anchor \ boxes$ 会得到更高的 $Avg \ IoU$ 。这表明使用 $k-means$ 生成 $anchor \ boxes$ 可以更好地表示模型并使其更容易学习。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:31]:
```suggestion
&emsp;表1中，我们将聚类策略得到的 $anchor \ boxes$ 和手工选取的 $anchor \ boxes$ 在最接近的 $Avg \ IoU$ 上进行比较。通过聚类策略得到的仅5种 $anchor \ boxes$ 中心的 $Avg \ IoU$ 为61.0，其性能类似于9个通过网络学习的 $anchor \ boxes$ 的60.9 (*即Avg IoU已经达到了Faster RCNN的水平*)。 而且使用9种 $anchor \ boxes$ 会得到更高的 $Avg \ IoU$ 。这表明使用 $k-means$ 生成 $anchor \ boxes$ 可以更好地表示模型并使其更容易学习。
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+
+1. 手动设定簇的个数k，假设k=2；
+2. 在所有样本中随机选取k个样本作为簇的初始中心，如下图（random clusters）中两个黄色的小星星代表随机初始化的两个簇中心；
+3. 计算每个样本离每个簇中心的距离（这里以欧式距离为例），然后将样本划分到离它最近的簇中。如下图（step 0）用不同的颜色区分不同的簇；
+4. 更新簇的中心，计算每个簇中所有样本的均值（方法不唯一）作为新的簇中心。如下图（step 1）所示，两个黄色的小星星已经移动到对应簇的中心；
+5. 重复第3步到第4步直到簇中心不在变化或者簇中心变化很小满足给定终止条件。如下图（step2）所示，最终聚类结果。
+
+![image](https://user-images.githubusercontent.com/109639975/200206147-46531a06-5011-4020-ab7c-967ddf9c0df2.png)
+
+### 什么是bpr?
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:66]:
```suggestion
### 什么是BPR?
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+BPR（bpr best possible recall来源于论文: [FCOS](https://arxiv.org/abs/1904.01355).
+
+原论文解释：
+
+> BPR is defined as the ratio of the number of ground-truth boxes a detector can recall at the most divided by all ground-truth boxes. A ground-truth box is considered being recalled if the box is assigned to at least one sample (i.e., a location in FCOS or an anchor box in anchor-based detectors) during training.
+
+&emsp;bpr(best possible recall): 最多能被召回的gt框数量 / 所有gt框数量 最大值为1 越大越好 小于0.98就需要使用k-means + 遗传进化算法选择出与数据集更匹配的anchors框。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:73]:
```suggestion
&emsp;bpr(best possible recall): 最多能被召回的ground truth框数量 / 所有ground truth框数量 最大值为1 越大越好 小于0.98就需要使用k-means + 遗传进化算法选择出与数据集更匹配的anchor boxes框。
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+
+### 什么是白化操作whiten？
+&emsp;白化的目的是去除输入数据的冗余信息。假设训练数据是图像，由于图像中相邻像素之间具有很强的相关性，所以用于训练时输入是冗余的；白化的目的就是降低输入的冗余性。
+
+输入数据集X，经过白化处理后，新的数据X’满足两个性质：
+
+1. 特征之间相关性较低；
+2. 所有特征具有相同的方差=1
+
+&emsp;常见的作法是：对每一个数据做一个标准差归一化处理（除以标准差）。scipy.cluster.vq.kmeans() 函数输入的数据就是必须是白化后的数据。相应的输出的[anchor](https://so.csdn.net/so/search?q=anchor&spm=1001.2101.3001.7020) k也是白化后的anchor，所以需要将anchor k 都乘以标准差恢复。
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:83]:
```suggestion
&emsp;常见的作法是：对每一个数据做一个标准差归一化处理（除以标准差）。scipy.cluster.vq.kmeans() 函数输入的数据就是必须是白化后的数据。相应的输出的[anchor](https://so.csdn.net/so/search?q=anchor&spm=1001.2101.3001.7020) k也是白化后的anchor，所以需要将anchor k 都乘以标准差恢复到原始数据的尺度。
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+
+```python
+def check_anchor_order(m):
+    """用在check_anchors最后 确定anchors和stride的顺序是一致的
+    Check anchor order against stride order for YOLOv5 Detect() module m, and correct if necessary
+    :params m: model中的最后一层 Detect层
+    """
+    # Check anchor order against stride order for YOLOv5 Detect() module m, and correct if necessary
+    # 计算anchor的面积 anchor area [9]
+    a = m.anchors.prod(-1).mean(-1).view(-1)  # mean anchor area per output layer
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:116]: 这一行很不好理解，建议用真实数据补充一下维度变化信息，

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+     # 计算最大anchor与最小anchor面积差
+    da = a[-1] - a[0]  # delta a
+    # 计算最大stride与最小stride差
+    ds = m.stride[-1] - m.stride[0]  # delta s
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:120]: 这里也可以补充一下m.stride是什么

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+## 3. kmean_anchors
+&emsp;这个函数才是这个这个文件的核心函数，功能：使用K-means + 遗传算法 算出更符合当前数据集的anchors。
+
+&emsp;这里不仅仅使用了k-means聚类，还使用了Genetic Algorithm遗传算法，在k-means聚类的结果上进行mutation变异。接下来简单介绍下代码流程：
+
+1. 载入数据集，得到数据集中所有数据的wh
+2. 将每张图片中wh的最大值等比例缩放到指定大小img_size，较小边也相应缩放
+3. 将bboxes从相对坐标改成绝对坐标（乘以缩放后的wh） 
+4. 筛选bboxes，保留wh都大于等于两个像素的bboxes
+5. 使用k-means聚类得到n个anchors（掉k-means包 涉及一个白化操作）
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:136]:
```suggestion
5. 使用k-means聚类得到n个anchors（调用k-means包 涉及一个白化操作）
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+        :params k: anchor框
+        :params wh: 整个数据集的wh [N, 2]
+        :return x: [N, 9] N个gt框与所有anchor框的宽比或高比(两者之中较小者)
+        :return x.max(1)[0]: [N] N个gt框与所有anchor框中的最大宽比或高比(两者之中较小者)
+        """
+        # [N, 1, 2] / [1, 9, 2] = [N, 9, 2]  N个gt_wh和9个anchor的k_wh宽比和高比
+        # 两者的重合程度越高 就越趋近于1 远离1(<1 或 >1)重合程度都越低
+        r = wh[:, None] / k[None]
+        # r=gt_height/anchor_height  gt_width / anchor_width  有可能大于1，也可能小于等于1
+        # torch.min(r, 1. / r): [N, 9, 2] 将所有的宽比和高比统一到<=1
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:176]:
```suggestion
        # flow.min(r, 1. / r): [N, 9, 2] 将所有的宽比和高比统一到<=1
```

File: docs/source_code_interpretation/utils/autoanchor_py.md [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+        # .min(2): value=[N, 9] 选出每个gt个和anchor的宽比和高比最小的值   index: [N, 9] 这个最小值是宽比(0)还是高比(1)
+        # [0] 返回value [N, 9] 每个gt个和anchor的宽比和高比最小的值 就是所有gt与anchor重合程度最低的
+        x = flow.min(r, 1. / r).min(2)[0]  # ratio metric
+        # x = wh_iou(wh, torch.tensor(k))  # IoU metric
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/autoanchor_py.md:180]:
```suggestion
        # x = wh_iou(wh, flow.tensor(k))  # IoU metric
```

**Review by BBuf — APPROVED** [review]
图片记得删除，上述commnet都需要修改
